gcp-client/src/main/java/com/google/graphite/platforms/plugin/client/ComputeClient.java

```diff
@@ -0,0 +1,541 @@
+   * @throws IOException If an error occurred in deleting the snapshot.
+   */
+  public void deleteSnapshot(String projectId, String snapshotName) throws IOException {
+    compute.snapshots().delete(projectId, snapshotName).execute();
+  }
+
+  /**
+   * Returns snapshot with name snapshotName
+   *
+   * @param projectId Google cloud project id.
+   * @param snapshotName Name of the snapshot to get.
+   * @return Snapshot object with given snapshotName. Null if not found.
+   * @throws IOException If an error occurred in retrieving the snapshot.
+   */
+  public Snapshot getSnapshot(String projectId, String snapshotName) throws IOException {
+    SnapshotList response;
+    Compute.Snapshots.List request = compute.snapshots().list(projectId);
+
+    do {
+      response = request.execute();
+      if (response.getItems() == null) {
+        continue;
+      }
+      for (Snapshot snapshot : response.getItems()) {
+        if (snapshotName.equals(snapshot.getName())) return snapshot;
+      }
+      request.setPageToken(response.getNextPageToken());
+    } while (response.getNextPageToken() != null);
+
+    return null;
+  }
+
+  /**
+   * Appends metadata to an instance. Any metadata items with existing keys will be overwritten.
+   * Otherwise, metadata is preserved. This method blocks until the operation completes.
+   *
+   * @param projectId
+   * @param zone
+   * @param instanceId
+   * @param items
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public Operation.Error appendInstanceMetadata(
+      String projectId, String zone, String instanceId, List<Metadata.Items> items)
+      throws IOException, InterruptedException {
+    zone = nameFromSelfLink(zone);
+    Instance instance = getInstance(projectId, zone, instanceId);
+    Metadata existingMetadata = instance.getMetadata();
+
+    List<Metadata.Items> newMetadataItems = mergeMetadataItems(items, existingMetadata.getItems());
+    existingMetadata.setItems(newMetadataItems);
+
+    Operation op =
+        compute.instances().setMetadata(projectId, zone, instanceId, existingMetadata).execute();
+    return waitForOperationCompletion(projectId, op.getName(), op.getZone(), 60 * 1000);
+  }
+
+  /**
+   * Blocks until an existing operation completes.
+   *
+   * @param projectId
+   * @param operationId
+   * @param zone
+   * @param timeout
+   * @return
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public Operation.Error waitForOperationCompletion(
+      String projectId, String operationId, String zone, long timeout)
+      throws IOException, InterruptedException {
+    if (Strings.isNullOrEmpty(operationId)) {
+      throw new IllegalArgumentException("Operation ID can not be null");
+    }
+    if (Strings.isNullOrEmpty(zone)) {
+      throw new IllegalArgumentException("Zone can not be null");
+    }
+    if (zone != null) {
+      String[] bits = zone.split("/");
+      zone = bits[bits.length - 1];
```

Comment:
Recommend verifying array length and throwing an exception with an informative message. Much easier to debug than an ArrayOutOfBoundIndex.